harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -701,6 +701,22 @@ def _train_with_boundaries(
         for batch_idx, batch in training_enumerator:
             epoch_idx, batch_in_epoch_idx = divmod(batch_idx, epoch_len)
 
+            # This can occur when we have already trained more than our trial
+            # will train for. For example, if we are provided a steps_completed
+            # larger than our searcher max length.
+            #
+            # TODO this is really ugly, I think this is the earliest this can be
+            # since we need the dataloader to be initialized to tell us what batch we are on?
+            # It also seems we should do the check before we actually train any batches? So the
+            # natural place where we check `should_stop` feels too late?
+            for step in train_boundaries:
```

Comment:
you just need the checkpoint state to be loaded to get `steps_completed` right? 

can this check just go in `_train_for_op` along with the `while not searcher_complete` line?

Reply:
Tried moving the check in ```_train_for_op```, it should be PRed

We still need to iterate through train_boundaries to check against each step right?

We would need to duplicate the epoch length logic to get epoch_idx from ```start_from_batch```?

```
        epoch_len = self.context._epoch_len
        assert epoch_len, "Training dataloader uninitialized."
        epoch_idx = divmod(self.start_from_batch, epoch_len)
```

Reply:
1. you don't need `start_from_batch`, that value exists because we need to initialize dataloaders before loading state. once state is loaded, it contains `batches_trained` and `epochs_trained` so you don't need to do the conversion. at the point when training starts, state should already be loaded so you shouldn't need to rely on `start_from_batch`. 
2. you'd need the `op.length` to get the `max_length` we're supposed to train for. but ideally you want it in `TrainUnit` form because there's already a lot of helper utils built around epoch/batch step logic. 

so, i think you can either:

change the beginning of `_train_for_op` to 

```
        searcher_length = TrainUnit._from_searcher_unit(
            op.length, self.searcher_unit, self.global_batch_size
        )
        while self._steps_until_complete(searcher_length) > 0:
```

or put a similar check in the caller of `_train_for_op` to opt out of training for the op altogether
```
assert train_unit

if self._steps_until_complete(train_unit=train_unit) < 1:
        break

self._train_for_op(...
```

Reply:
actually, i think it's possible that we continue from a checkpoint and the op has finished training but it didn't validate? so we'd probably still want this part outside the while loop in `_train_for_op`

```
# Finished training for op. Perform final checkpoint/validation if necessary.
        if not self._validation_is_current():
            self._validate(op)
```

so probably don't do the second option i mentioned. 😄

Reply:
Thanks for the suggestions, added your first suggestion and seems to be working. 

I did have to update the ```assert op._completed``` to be a ```break```